kythe/go/serving/pipeline/beam.go

```diff
@@ -624,3 +631,91 @@ func groupEdges(src *spb.VName, nodeStream func(**ppb.Node) bool, edgeStream, re
 
 	emitSet("edgeSets:"+set.Source.Ticket, set)
 }
+
+// Documents returns a Kythe documentation table derived from the Kythe input
+// graph.  The beam.PCollection has elements of type KV<string,
+// *srvpb.Document>.
+func (k *KytheBeam) Documents() beam.PCollection {
+	s := k.s.Scope("Documents")
+
+	docs := beam.Seq(s, k.nodes, &nodes.Filter{
+		FilterByKind: []string{kinds.Doc},
+		IncludeFacts: []string{facts.Text},
+		IncludeEdges: []string{edges.Documents},
+	}, nodeToDocs)
+	markedSources := beam.Seq(s, k.nodes, &nodes.Filter{
+		IncludeFacts: []string{facts.Code},
+		IncludeEdges: []string{},
+	}, parseMarkedSource)
+	children := beam.Seq(s, k.nodes, &nodes.Filter{
+		IncludeFacts: []string{},
+		IncludeEdges: []string{edges.ChildOf},
+	}, nodeToChildren)
+
+	return beam.ParDo(s, completeDocument, beam.CoGroupByKey(s, markedSources, docs, children))
+}
+
+// completeDocument emits a single *srvpb.Document per *spb.VName source.
+func completeDocument(key *spb.VName, msStream func(**cpb.MarkedSource) bool, docStream func(**srvpb.Document) bool, childStream func(**spb.VName) bool, emit func(string, *srvpb.Document)) {
+	var ms *cpb.MarkedSource
+	msStream(&ms)
+
+	var doc *srvpb.Document
+	if !docStream(&doc) {
```

Comment:
consider doing this first, so that you don't bother populating `ms` if you won't be using it

Reply:
✓